custom/icds_reports/sqldata/exports/poshan_progress_report.py

```diff
@@ -0,0 +1,188 @@
+    "% of trimester three women counselled on immediate and EBF",
+    "Height Measurement Efficiency",
+    "% of children between 6 months -3 years, P&LW provided THR for atleast 21+ days",
+    "% of children between 3-6 years provided SNP for atleast 21+ days"]
+
+COLS_SUMMARY = [
+    'state_name', 'district_name', 'num_launched_districts', 'num_launched_blocks', 'num_launched_awcs',
+    'avg_days_awc_open_percent', 'visits_percent', 'pse_attended_21_days_percent', 'weighed_percent',
+    'counsel_immediate_bf_percent', 'height_measured_in_month_percent', 'thr_percent',
+    'lunch_count_21_days_percent'
+]
+
+
+class PoshanProgressReport(object):
+    title = 'Poshan Progress Report'
+
+    def __init__(self, config, loc_level=0, beta=False, show_test=False):
+        self.config = config
+        self.loc_level = loc_level
+        self.beta = beta
+        self.show_test = show_test
+        self.layout = self.config['report_layout']
+        if self.config['data_format'] == 1:
+            self.report_type = 'month'
+        else:
+            self.report_type = 'quarter'
+            self.quarter = self.config['quarter']
+            self.year = self.config['year']
+
+    def _generate_quarter_years(self):
+        months = []
+        end_month = self.quarter * 3
+        for i in range(end_month - 2, end_month + 1):
+            months.append(date(self.year, i, 1))
+        return months
+
+    def structure_data(self, headers, cols, rows):
+        """
+        :param headers: headers of the excel file
+        :param cols: columns mapped with headers
+        :param rows: row data
+        :return: list of list contain the excel rows
+        """
+        excel_rows = [headers]
+        for row in rows:
+            row_data = []
+            for col in cols:
+                row_data.append(row[col])
+            excel_rows.append(row_data)
+        return excel_rows
+
+    def quarter_wise(self, filters, order_by):
+        """
+        :param filters: quaterwise filter (months in [month array])
+        :param order_by: order by columns
+        :return: excel_rows
+        """
+        query_set = PoshanProgressReportView.objects.filter(**filters).order_by(*order_by)
+        cols = COLS_COMPREHENSIVE[:]
+        cols.append('district_id')  # used as key for the dict
+        data = query_set.values(*cols)
+        row_data_dict = {}
+        dummy_row = ['', ''] + [0 for _ in range(0, len(COLS_COMPREHENSIVE) - 2)]
+        headers = HEADERS_COMPREHENSIVE
+        # generate a dict
+        # {'disitrict_id': [contains the excel row with sum of col values for all months eg. m1+m2+m3]}
+        for row in data:
+            row_data = dummy_row
+            for k, v in row.items():
+                if k == 'state_name' or k == 'district_name':
+                    row_data[COLS_COMPREHENSIVE.index(k)] = v
+                elif k != 'district_id':
+                    row_data[COLS_COMPREHENSIVE.index(k)] += v
```

Comment:
could you please explain how are we ensuring the location uniqueness summation here. i.e We are not mistakenly summing x of s1 state with x of s2 state?

Reply:
fixed